KomaMRIFiles/src/Sequence/WritePulseq.jl

```diff
@@ -0,0 +1,407 @@
+Requires an initial integer counter "id_shape_cnt" to asign IDs incrementally.
+"""
+function get_rfunique(rfunique_obj_id::Vector, id_shape_cnt::Integer, seq::Sequence)
+    # Find the unique shapes (magnitude, phase and time shapes) and assign IDs
+    rfunique_abs_id, rfunique_ang_id, rfunique_tim_id = [], [], []
+    for (obj,_) ∈ rfunique_obj_id
+        shape_abs = abs.(obj.A) / maximum(abs.(obj.A))
+        if all([!(length(shape_abs) == length(shape_abs_unique) && shape_abs ≈ shape_abs_unique) for (shape_abs_unique,_) ∈ rfunique_abs_id])
+            push!(rfunique_abs_id, [shape_abs, id_shape_cnt])
+            id_shape_cnt += 1
+        end
+        shape_ang = -angle.(obj.A)/2π  # [-0.5, 0.5]
+        if all([!(length(shape_ang) == length(shape_ang_unique) && all(isapprox.(unwrap_angle.(shape_ang .- shape_ang[1]), unwrap_angle.(shape_ang_unique .- shape_ang_unique[1]), atol=1e-7))) for (shape_ang_unique,_) ∈ rfunique_ang_id])
+            push!(rfunique_ang_id, [unwrap_negative_angle.(shape_ang), id_shape_cnt])   # this always garanties non-negative shaped angles
+            id_shape_cnt += 1
+        end
+        if isa(obj.T, Vector{<:Number})
+            shape_tim = cumsum([0; obj.T]) / seq.DEF["RadiofrequencyRasterTime"]
+            if all([!(length(shape_tim) == length(shape_tim_unique) && shape_tim ≈ shape_tim_unique) for (shape_tim_unique,_) ∈ rfunique_tim_id])
+                push!(rfunique_tim_id, [shape_tim, id_shape_cnt])
+                id_shape_cnt += 1
+            end
+        end
+    end
+    return rfunique_abs_id, rfunique_ang_id, rfunique_tim_id, id_shape_cnt
+end
+
+
+"""
+    gradunique_amp_id, gradunique_tim_id, id_shape_cnt = get_gradunique(gradunique_obj_id::Vector, id_shape_cnt::Integer, seq::Sequence)
+
+Returns the unique shapes for the amplitude and time of the "gradunique_obj_id" vector.
+Requires an initial integer counter "id_shape_cnt" to asign IDs incrementally.
+"""
+function get_gradunique(gradunique_obj_id::Vector, id_shape_cnt::Integer, seq::Sequence)
+    # Find shapes for magnitude and time gradients
+    gradunique_amp_id, gradunique_tim_id = [], []
+    for (obj,_) ∈ gradunique_obj_id
+        shape_amp = obj.A / magsign(obj.A)
+        if all([!(length(shape_amp) == length(shape_amp_unique) && shape_amp ≈ shape_amp_unique) for (shape_amp_unique,_) ∈ gradunique_amp_id])
+            push!(gradunique_amp_id, [shape_amp, id_shape_cnt])
+            id_shape_cnt = id_shape_cnt + 1
+        end
+        if isa(obj.T, Vector{<:Number})
+            shape_tim = cumsum([0; obj.T])/seq.DEF["GradientRasterTime"]
+            if all([!(length(shape_tim) == length(shape_tim_unique) && shape_tim ≈ shape_tim_unique) for (shape_tim_unique,_) ∈ gradunique_tim_id])
+                push!(gradunique_tim_id, [shape_tim, id_shape_cnt])
+                id_shape_cnt = id_shape_cnt + 1
+            end
+        end
+    end
+    return gradunique_amp_id, gradunique_tim_id, id_shape_cnt
```

Comment:
All of these comparisons are not defined in isapprox? what is the purpose of them then

Reply:
They are different from isapprox (for Koma events), and these new comparisons are necessary.

For instance, let's consider two different Grad events that only differ their amplitudes by a multiplicative factor. They should have the same amp_id. The isapprox function cannot identify this.